### DISCUSSION TITLE
Solving the read-only locals and parameters problem: read-only classes and members

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#7367] [Language Ideas] [answered]
A very popular and long running feature request thread is for C# to support read-only locals and parameters (#188). This has spawned a mass of suggestions along with objections from the language team. Those objections tend to centre around three key ideas:

1. The scope of locals and parameters is so small, there's little real-world benefit to marking them read-only
2. There is reluctance to have "best practice" C# be full of `readonly` keywords
3. There is a desire to have a new dialect of C# where variables are read-only by default 

These objections effectively killed off the discussion ... until non-record primary constructors landed. The scope of their parameters is class/struct-wide. There is a much stronger need for being able to mark them read-only than there is for normal constructors and (other) methods. This has both prompted the team to consider read-only parameters for this specific case (https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-07-17.md), but also re-opened the discussion on read-only locals and parameters in general. 

Should the language just allow everything to be marked with `readonly` (with maybe simple local declarations using `let`):
```cs
public class C 
{
    public void M(readonly object o) 
    {
        let w = o;
        let x = (readonly p) => p;
        
        if (o is readonly string y)
        {
            ...
        }
        else if (TryGetAnotherWay(out readonly var z)
        {
            ...
        }
    }
}
```
Even with `let`, that's a lot of `readonly`'s. Maybe `let` could be used for out-parameters and patterns though:
 ```cs
 public class C 
{
    public void M(readonly object o) 
    {
        let w = o;
        let x = (readonly p) => p;

        if (o is let string y)
        {
            ...
        }
        else if (TryGetAnotherWay(out let z)
        {
            ...
        }
    }
}
```
I don't know what others think, but `o is let string y` and `out let z` are not selling themselves to me as a good idea.

Another suggestion has been to using an analyzer to enforce read-only variables and attributes to override this. This would require attributes to be supported in odd places though:
```cs
public class C 
{
    public void M([Mutable] object o) 
    {
        [Mutable]var x = o;
        var x = ([Mutable] p) => p;

        if (o is [Mutable]string y)
        {
            ...
        }
        else if (TryGetAnotherWay(out [Mutable]var z)
        {
            ...
        }
    }
}
```
and one last suggestion is that the compiler support a special immutable-by-default dialect:
```cs
# immutable enable
public class C 
{
    public void M(mutable object o) 
    {
        mutable var x = o;
        var x = (mutable p) => p;

        if (o is mutable string y)
        {
            ...
        }
        else if (TryGetAnotherWay(out mutable var z)
        {
            ...
        }
    }
}
```
Nullable reference types introduced such a dialect. But the gains there were huge and the syntactic impact was small. The opposite is true here though. 

There is an alternative that has been missed here: allowing the whole class to be marked `readonly`:
```cs
// w, x, y and z are made readonly by the class
public readonly class C 
{
    public void M(object o)
    {
        var w = o;
        var x = (p) => p;
        
        if (o is string y)
        {
            ...
        }
        else if (TryGetAnotherWay(out var z)
        {
            ...
        }
    }
}
```
The "dialect" effect is achieved by marking the class as readonly. Everything then becomes read-only in that class. Alternatively, if most of the class needs to be read-only, then the members can be marked - just like with readonly structs - except for those methods that need to mutate locals. 

This approach addresses all three of the team's concerns: the read-onliness is scoped to the whole class or method via one `readonly`, "best practice" C# doesn't become full of `readonly` keywords everywhere and a new dialect is avoided. And one of the most demanded features is finally delivered. Everyone's a winner...

**Comment by HaloFour** [comment, marked as answer]
Except that `readonly` is already a valid modifier on `struct` and on methods of a `struct`, and neither has this behavior:

```cs
public readonly struct S {
    public readonly int X; // readonly modifier required
    
    public S(int x) {
        x = 123; // legal
        this.X = x;
    }
    
    public readonly void M(int x) {
        x = 123; // legal
        if (TryGetAnother(out int y)) {
            y = 789; // also legal
        }
    }
    
    private static bool TryGetAnother(out int y) {
        y = 456;
        return true;
    }
}
```

**Reply by DavidArno** [reply]
Seriously, variables in a readonly method in a readonly struct are not readonly?!???

I give up. C#'s approach to mutability is insane 😭

**Reply by CyrusNajmabadi** [reply]
Of course not.  Same as how variables in a const method in c++ are not const.

**Reply by HaloFour** [reply]
Yeah, the purpose of the `readonly` method is to enforce that the state of the struct itself isn't mutated and to advertise that to the caller so that it knows that it doesn't have to make a defensive copy.  It doesn't apply any opinions beyond that.

Also, #3885  😅

**Reply by CyrusNajmabadi** [reply]
'readonly' for a member means "can safely be called (no need for a copy) on a readonly instance of a struct". In other words, being readonly means it can't mutate state of the instance.  Locals are not state. They're the temporary values created to compute values. 

Imagine how useless it would be if all your temporaries were readonly.  You wouldn't be able to even have a simple `for` loop anymore.

**Reply by HaloFour** [reply]
> You wouldn't be able to even have a simple `for` loop anymore.

Well, not a C-style `for` loop.

```scala
for (x <- 0 to 10) {
    // x is immutable
}

var y: Int;
for (y <- 0 to 10) {
    // y is mutable
}
```

Not at all different to how C# treats foreach iteration variables, really.

**Reply by CyrusNajmabadi** [reply]
>  C#'s approach to mutability is insane 

I think you're confusing two things.  C# cares heavily about mutability *when it comes to instance state*. That's why we have readonly state.  It's why copies of structs happen (to not mutate state).   It's why members can be marked readonly *to indicate they don't assign to state*. 

This is why, for example, we're so willing to open the discussion about PC parameters and 'readonly'. Because they are state, and ensuring you know your state won't change *is* a way of substantively reducing real world bugs. 

But state of instances is not at all the same as the local, (literally temporary) scratch space used during computation to compute values.  In that domain, c# remains firmly in the "meh, it's entirely fine for temps to be assigned to".  One of the clearest examples of that is the simple 'for' loop.  Its primary use case is exactly to mutate a local as that is so high value to the user. 

Unlike instance state, we also do not see any sort of widespread, systemic issues arising from local state being mutable.  As such, there is no pressing need to make it possible to mark locals as being non-assignable. 

So our approach to mutability is very clear and sane: state is good to be able to make readonly. Locals are practically irrelevant. 

And then language has been designed in a corresponding fashion.

**Reply by CyrusNajmabadi** [reply]
> Well, not a C-style for loop.

Correct. But that would be a problem since basic for-loops are so present and idiomatic.  I'm using that just as an example of how people (even with readonly state) will quite normally have mutable locals as the two domains have wildly different needs, and the data indicates that mutability in one is a serious issue for correctness, while mutability in the other is not.

**Reply by HaloFour** [reply]
@CyrusNajmabadi 

> But state of instances is not at all the same as the local, (literally temporary) scratch space used during computation to compute values. In that domain, c# remains firmly in the "meh, it's entirely fine for temps to be assigned to". One of the clearest examples of that is the simple 'for' loop. Its primary use case is exactly to mutate a local as that is so high value to the user.

Not to sound like a broken record, but captures changed that calculus.  Now whether locals are scratch or state is a gray area determined by usage.  😀

**Reply by CyrusNajmabadi** [reply]
As I mentioned above:

> This is why, for example, we're so willing to open the discussion about PC parameters and 'readonly'. Because they are state, and ensuring you know your state won't change is a way of substantively reducing real world bugs.

Indeed, we've been very empathetic on this point since the first feedback came in. That this is a space we feel is def one we will iterate on since it overlaps this critical aspect of type development. 

Now, as to what the resultant iteration will be, I cannot say. Good arguments have been made in many directions about this. Personally, I'm ok with the defaults changing for PC params, *or* the ability to add 'readonly' to them. My preference is actually the latter as I both think it's clearer, and it allows a more consistent 'glow up' story to allowing field declarations with the param. But I'm also ok with implicit readonly-ness, where the user has to define a field manually if they don't want that.
